### DISCUSSION TITLE
Working with large local files

### DISCUSSION THREAD

**Discussion by nebfield** [pyodide/pyodide#4371] [Q&A] [answered]
I'm building a [simple test application](https://github.com/ebi-gdp/crypt4gh-wasm) with pyodide. I'd like to read fairly large files from a local directory and encrypt them with a web worker.

My test app works well (and fast!) thanks to the excellent docs. I basically:

1. Select a directory and get permission to mount with `mountNativeFS`
2. Select a file to read in the mounted directory using a file picker
3. Get a web worker to run encryption in python using the mounted file path. The encryption process buffers data, encrypts it, and writes to a new file in the mounted directory
4. Synchronise the file system after encryption runs

But I'm running into an issue with files larger than a few gigabytes. These large files don't appear in the mounted directory. 

I naively assumed reading large files wouldn't be a problem with the File System Access API and `mountNativeFS`, but I can see there are some limitations that weren't apparent to me from reading the docs: https://github.com/pyodide/pyodide/issues/3881#issuecomment-1562579789 

Are there any reasonable ways to work around this limit? Could it be possible to stream data from a large local file to the python process?

I've seen an [example provided here](https://stackoverflow.com/a/75871580). So I'm thinking a sensible solution might be to stream a file with JS, pass to the python worker, and then write a stream of bytes to disk with JS.

**Comment by ryanking13** [comment, marked as answer]
> I naively assumed reading large files wouldn't be a problem with the File System Access API and mountNativeFS, but I can see there are some limitations that weren't apparent to me from reading the docs: https://github.com/pyodide/pyodide/issues/3881#issuecomment-1562579789

Yeah, `mountNativeFS` also has the same memory size limitation, as it stores the data in the memory when loading / storing the data from local file system.

> Are there any reasonable ways to work around this limit? Could it be possible to stream data from a large local file to the python process? I've seen an [example provided here](https://stackoverflow.com/a/75871580). So I'm thinking a sensible solution might be to stream a file with JS, pass to the python worker, and then write a stream of bytes to disk with JS.

Yeah, I didn't try, but I think it is a reasonable approach. I don't think file streaming will work with the current Emscripten file system, including the native file system (especially regarding the fact that it is asynchronous), so I think you need to deal with it at the JS level.

**Reply by nebfield** [reply]
Thanks for your help 😄
